x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java

```diff
@@ -0,0 +1,260 @@
+        "/org/elasticsearch/xpack/idp/saml/support/saml-schema-assertion-2.0.xsd",
+        "/org/elasticsearch/xpack/idp/saml/support/xenc-schema.xsd",
+        "/org/elasticsearch/xpack/idp/saml/support/xmldsig-core-schema.xsd"};
+
+    private static final ThreadLocal<DocumentBuilder> THREAD_LOCAL_DOCUMENT_BUILDER = ThreadLocal.withInitial(() -> {
+        try {
+            return SamlUtils.getHardenedBuilder(XSD_FILES);
+        } catch (Exception e) {
+            throw new ElasticsearchSecurityException("Could not load XSD schema file", e);
+        }
+    });
+
+    public SamlAuthnRequestValidator(SamlIdentityProvider idp) {
+        SamlUtils.initialize();
+        this.idp = idp;
+    }
+
+    public SamlValidateAuthnRequestResponse processQueryString(String queryString) {
+        final List<String> parameterList = Arrays.asList(queryString.split("&"));
+        final Map<String, String> parameters = new HashMap<>();
+        if (parameterList.isEmpty()) {
+            throw new IllegalArgumentException("Invalid Authentication Request query string");
+        }
+        RestUtils.decodeQueryString(queryString, 0, parameters);
+        logger.trace(new ParameterizedMessage("Parsed the following parameters from the query string: {}", parameters));
+        // We either expect at least a single parameter named SAMLRequest or at least 3 ( SAMLRequest, SigAlg, Signature )
+        final String samlRequest = parameters.get("SAMLRequest");
+        final String relayState = parameters.get("RelayState");
+        final String sigAlg = parameters.get("SigAlg");
+        final String signature = parameters.get("Signature");
+        if (null == samlRequest) {
+            logAndThrow(() -> new ParameterizedMessage("Query string [{}] does not contain a SAMLRequest parameter", queryString));
+        }
+        AuthnRequest authnRequest = null;
+        // We consciously parse the AuthnRequest before we validate its signature as we need to get the Issuer, in order to
+        // verify if we know of this SP and get its credentials for signature verification
+        final Element root = parseSamlMessage(inflate(decodeBase64(samlRequest)));
+        if ("AuthnRequest".equals(root.getLocalName()) == false
+            || "urn:oasis:names:tc:SAML:2.0:protocol".equals(root.getNamespaceURI()) == false) {
+            logAndThrow(() -> new ParameterizedMessage("SAML message [{}] is not an AuthnRequest", SamlUtils.toString(root, true)));
+        }
+        try {
+            authnRequest = SamlUtils.buildXmlObject(root, AuthnRequest.class);
+        } catch (Exception e) {
+            logAndThrow(() -> new ParameterizedMessage("Cannot process AuthnRequest [{}]", SamlUtils.toString(root, true)), e);
+        }
+        final SamlServiceProvider sp = getSpFromIssuer(authnRequest.getIssuer());
+        if (signature != null) {
+            if (sigAlg == null) {
```

Comment:
I can check the spec if you don't know, but should `sigAlg != null && signature == null` be an error too?

Reply:
Correct, it's either both or none